pkg/agent/attestor/attestor.go

```diff
@@ -0,0 +1,271 @@
+	ServerAddress *net.TCPAddr
+	NodeClient    node.NodeClient
+}
+
+type attestor struct {
+	c *Config
+}
+
+func New(config *Config) Attestor {
+	return &attestor{c: config}
+
+}
+
+func (a *attestor) Attest() (*AttestationResult, error) {
+	bundle, err := a.loadBundle()
+	if err != nil {
+		return nil, err
+	}
+	svid, key, err := a.loadSVID()
+	if err != nil {
+		return nil, err
+	}
+
+	if svid == nil {
+		svid, bundle, err = a.newSVID(key, bundle)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &AttestationResult{Bundle: bundle, SVID: svid, Key: key}, nil
+}
+
+func (a *attestor) loadSVID() (*x509.Certificate, *ecdsa.PrivateKey, error) {
+	mgrs := a.c.Catalog.KeyManagers()
+	if len(mgrs) > 1 {
+		return nil, nil, errors.New("more than one key manager configured")
+	}
+
+	mgr := mgrs[0]
+	fResp, err := mgr.FetchPrivateKey(&keymanager.FetchPrivateKeyRequest{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("load private key: %v", err)
+	}
+
+	svid := a.readSVIDFromDisk()
+	if len(fResp.PrivateKey) > 0 && svid == nil {
+		a.c.Log.Warn("Private key recovered, but no SVID found")
+	}
+
+	var keyData []byte
+	if len(fResp.PrivateKey) > 0 && svid != nil {
+		keyData = fResp.PrivateKey
+	} else {
+		gResp, err := mgr.GenerateKeyPair(&keymanager.GenerateKeyPairRequest{})
+		if err != nil {
+			return nil, nil, fmt.Errorf("generate key pair: %s", err)
+		}
+
+		svid = nil
+		keyData = gResp.PrivateKey
+	}
+
+	key, err := x509.ParseECPrivateKey(keyData)
+	if err != nil {
+		return nil, nil, fmt.Errorf("parse key from keymanager: %v", key)
+	}
+
+	return svid, key, nil
+}
+
+func (a *attestor) loadBundle() ([]*x509.Certificate, error) {
+	bundle, err := manager.ReadBundle(a.bundleCachePath())
+	if err == manager.ErrNotCached {
+		bundle = a.c.TrustBundle
+	} else if err != nil {
+		return nil, err
+	}
+
+	if bundle == nil {
+		return nil, errors.New("load bundle: no bundle provided")
+	}
+
+	if len(bundle) < 1 {
+		return nil, errors.New("load bundle: no certs in bundle")
+	}
+
+	return bundle, nil
+
+}
+
+func (a *attestor) attestationData() (*nodeattestor.FetchAttestationDataResponse, error) {
+
+	resp := &nodeattestor.FetchAttestationDataResponse{}
+
+	if a.c.JoinToken != "" {
+		data := &common.AttestedData{
+			Type: "join_token",
+			Data: []byte(a.c.JoinToken),
+		}
+
+		id := &url.URL{
+			Scheme: "spiffe",
+			Host:   a.c.TrustDomain.Host,
+			Path:   path.Join("spire", "agent", "join_token", a.c.JoinToken),
+		}
+
+		resp.AttestedData = data
+		resp.SpiffeId = id.String()
+		return resp, nil
+	}
+
+	plugins := a.c.Catalog.NodeAttestors()
+	if len(plugins) > 1 {
+		return nil, errors.New("more then one node attestor configured")
+	}
+	attestor := plugins[0]
+
+	return attestor.FetchAttestationData(&nodeattestor.FetchAttestationDataRequest{})
+}
+
+func (a *attestor) bundleCachePath() string {
+	return path.Join(a.c.DataDir, "bundle.der")
+}
+
+func (a *attestor) agentSVIDPath() string {
+	return path.Join(a.c.DataDir, "agent_svid.der")
```

Comment:
Same comment here. This path is also used in https://github.com/spiffe/spire/pull/392/files#diff-64c23753751bef052315a093bb62faeaR175